### PR TITLE
Updated starknet-verify action

### PR DESCRIPTION
**Pull request by dribeiro-ShardLabs**
Adapt the contract verification to be compatible with the new Voyager update

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -201,7 +201,13 @@ extendEnvironment((hre) => {
 
 task("starknet-verify", "Verifies the contract in the Starknet network.")
     .addOptionalParam("starknetNetwork", "The network version to be used (e.g. alpha)")
-    .addParam("path", "The path of the cairo contract (e.g. contracts/conract.cairo)")
+    .addOptionalParam("path", "The path of the cairo contract (e.g. contracts/conract.cairo)")
+    .addOptionalParam(
```
Comment (FabijanC) [inline src/index.ts:205]: Why have both `path` and `paths`? Shouldn't it be just `path` which is a variadic parameter? Also that way users wouldn't have to quote their space separated paths.

Another approach could be to keep the `path` parameter and introduce a `dependencies` optional variadic parameter.

File: test/general-tests/starknet-verify/check.sh [hunk 2]
```diff
@@ -4,13 +4,24 @@ set -e
 echo "The starknet-verify test is temporarily disabled."
 echo "To enable it back, uncomment the lines in its check.sh."
 
-# npx hardhat starknet-compile contracts/util.cairo
+
+# npx hardhat starknet-compile contracts/contract.cairo contracts/util.cairo
```
Comment (FabijanC) [inline test/general-tests/starknet-verify/check.sh:8]: It should be commented here why the util contract is compiled as well.

File: test/general-tests/starknet-verify/check.sh [hunk 3]
```diff
@@ -4,13 +4,24 @@ set -e
...
 # output=$(npx hardhat starknet-deploy --starknet-network "$NETWORK" starknet-artifacts/contracts/util.cairo/)
-# echo $output
-# address=$(echo $output | sed -r "s/.*Contract address: (\w*).*/\1/")
+# util_address=$(echo $output | sed -r "s/.*Contract address: (\w*).*/\1/")
+
+# output=$(npx hardhat starknet-deploy --starknet-network "$NETWORK" starknet-artifacts/contracts/contract.cairo/)
+# main_address=$(echo $output | sed -r "s/.*Contract address: (\w*).*/\1/")
 
-# echo "Address: $address"
+# echo "Util contract address: $util_address"
```
Comment (FabijanC) [inline test/general-tests/starknet-verify/check.sh:16]: You can `echo` this right after the deployment.

File: src/task-actions.ts [hunk 4]
```diff
@@ -291,34 +297,100 @@ export async function starknetVoyagerAction(args: TaskArguments, hre: HardhatRun
...
+    paths.forEach(function (item, index) {
+        if (!path.isAbsolute(item)) {
+            paths[index] = path.normalize(path.join(hre.config.paths.root, item));
+            if (!fs.existsSync(paths[index])) {
+                throw new HardhatPluginError(PLUGIN_NAME, `File ${paths[index]} does not exist`);
+            }
+        }
+    });
+
+    const bodyFormData = new URLSearchParams();
```
Comment (FabijanC) [inline src/task-actions.ts:323]: I find it unintuitive that for `bodyFormData` is constructed using `URLSearchParams`. Is this really the way?

File: src/task-actions.ts [hunk 5]
```diff
@@ -291,34 +297,100 @@ export async function starknetVoyagerAction(args: TaskArguments, hre: HardhatRun
...
+                throw new HardhatPluginError(PLUGIN_NAME, msg);
+            }
+            default: {
+                const msg = `Something went wrong when trying to verify the code at address ${address}`;
+                throw new HardhatPluginError(PLUGIN_NAME, msg);
+            }
+        }
+    });
+    console.log(`Contract has been successfuly verified at address ${address}`);
+    return;
```
Comment (FabijanC) [inline src/task-actions.ts:358]: why return?

File: src/task-actions.ts [hunk 6]
```diff
@@ -291,34 +297,100 @@ export async function starknetVoyagerAction(args: TaskArguments, hre: HardhatRun
...
-    let contractPath = args.path;
+    if (args.path) {
+        await handleSingleContractVerification(args.path, args.address, voyagerUrl, hre);
+    } else {
+        const splitPaths = args.paths.split(" ");
+        await handleMultiPartContractVerification(splitPaths, args.address, voyagerUrl, hre);
+    }
+}
+
+async function handleMultiPartContractVerification(
```
Comment (FabijanC) [inline src/task-actions.ts:308]: These two functions should be refactored. There is a lot of common functionality like path checking and response checking.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: src/task-actions.ts (hunk 4 above)
Comment (dribeiro-ShardLabs) [reply src/task-actions.ts:323]: the other option for the formData would be to add a new dependency 'form-data', but the URLSearchParams works exactly the same

File: src/index.ts [hunk 7]
```diff
@@ -203,6 +203,12 @@ task("starknet-verify", "Verifies the contract in the Starknet network.")
     .addOptionalParam("starknetNetwork", "The network version to be used (e.g. alpha)")
     .addParam("path", "The path of the cairo contract (e.g. contracts/conract.cairo)")
     .addParam("address", "The address where the contract is deployed")
+    .addOptionalVariadicPositionalParam(
+    "paths",
+    "The paths of the dependencies of the contract specified in --param" +
```
Comment (FabijanC) [inline src/index.ts:208]: --param? or --path?

File: src/index.ts [hunk 8]
```diff
@@ -203,6 +203,12 @@ task("starknet-verify", "Verifies the contract in the Starknet network.")
     .addOptionalParam("starknetNetwork", "The network version to be used (e.g. alpha)")
     .addParam("path", "The path of the cairo contract (e.g. contracts/conract.cairo)")
```
Comment (FabijanC) [inline src/index.ts:204]: Would adding the adjective "main" to "contract" be useful?

**Review by FabijanC — CHANGES_REQUESTED** [review]

**Review by FabijanC — APPROVED** [review]

**Pull request merged**